### DISCUSSION TITLE
Interactive mode for prqlc

### DISCUSSION THREAD

**Discussion by vanillajonathan** [PRQL/prql#3349] [Ideas]
`prqlc` could have a interactive mode though an `-i` or `--interactive` option.

Example:
```
$ prqlc --interactive
prqlc 0.9.3
Type '?' for help.

In [1:] from invoices
   ...: filter income > 5
   ...: sort {-sum_income}
   ...: take 20
   ...: 
SELECT
  *
FROM
  invoices
WHERE
  income > 5
ORDER BY
  sum_income DESC
LIMIT
  20

In [2:]
```

**Comment by eitsupi** [comment]
Wouldn't it be better to create another tool rather than making prqlc bigger?

**Reply by vanillajonathan** [reply]
Good point. It could be called `iprql` like `ipython`.

**Comment by max-sixty** [comment]
I think this is a great idea, and I was thinking of doing something like this at one point — maybe more like a TUI with two panes, like the playground. 

There could be shortcut keys for copying to clipboard, for viewing intermediate compiler outputs, etc.

The main reason _not_ to do it is that we already have the playground & VS Code. But the playground takes ~30s to start on local code each time. And VS Code works best when saving a `.prql` file, rather than just typing an example.

**Reply by vanillajonathan** [reply]
I was thinking something more REPL like `ipython`. Not a two-pane TUI like the playground.

**Reply by max-sixty** [reply]
What's the advantage of a REPL? I haven't thought about it that much, but seems like two main differences:
- The two-pane layout gives an instant result, vs. REPL gives a result on every evaluation
- The two-pane layout's state is completely defined by what's currently in the document; there's no history. A REPL has variables that can be created / overwritten. (which I think is also worse, though maybe folks have other preferences?)

**Reply by vanillajonathan** [reply]
Good point. Maybe a two-pane TUI would be a better fit for PRQL. When I made this post I got the idea from `ipython`.

**Reply by snth** [reply]
Didn't the original pyprql TUI do some of this?

**Reply by max-sixty** [reply]
Yes it did! It was great for this!

**Comment by snth** [comment]
So to some extent `prqlc` already does this:

```sh
❯ prqlc compile | duckdb
Enter PRQL, then press ctrl-d to compile:

from [{greet="hello @vanillajonathon"}]
┌────────────────────────┐
│         greet          │
│        varchar         │
├────────────────────────┤
│ hello @vanillajonathon │
└────────────────────────┘
```

However you don't get much interactivity beyond what's offered by your shell. IIRC there are readline wrappers that might give you basic functionality with 5 minutes of work. Something like https://github.com/hanslub42/rlwrap.

Otherwise using some of the https://charm.sh/ tools?

**Reply by snth** [reply]
This is very much in my wheelhouse and I have big plans to build something along these lines but I can't promise anything at this stage since I'm unlikely to keep any such promise.
